README.md
# SikuliX Cookbook

[![Cookbook Version](http://img.shields.io/cookbook/v/sikulix.svg?style=flat-square)][cookbook]
[![Build Status](http://img.shields.io/travis/dhoer/chef-sikulix.svg?style=flat-square)][travis]

[cookbook]: https://supermarket.chef.io/cookbooks/sikulix
[travis]: https://travis-ci.org/dhoer/chef-sikulix

Installs SikuliX 1.1.0 (http://www.sikulix.com/).

## Requirements

- Java (not installed by this cookbook)
- Chef 11+

### Platforms

- Mac OS X
- Ubuntu
- Windows

### Cookbooks

- apt
- windows

## Usage

Include the sikulix default recipe in your run list or recipe.

**IMPORTANT!** Since there are a variety of setup options to choose from, all option attributes default to `false`. You 
must select at least one setup option to install by setting its corresponding attribute to `true`.

Example

```ruby
"override_attributes": {
  "sikulix": {
    "setup": {
      "java_api": true
    }
  }
}
```

## Attributes

Pack1: I want SikuliX, containing the Sikuli IDE and allowing to run Sikuli scripts from commandline.

- `node['sikulix']['setup']['ide_jython']` - Python language level 2.7. Default `false`.
- `node['sikulix']['setup']['ide_jruby']` - Ruby language level 1.9 and 2.0. Default `false`.

Pack2: I want to develop in Java, Jython or other Java aware scripting language using NetBeans, Eclipse, or other IDE's.

- `node['sikulix']['setup']['java_api']` - Default `false`.  

I want to use the Tesseract based OCR features (You Should know what you are doing!).

- `node['sikulix']['setup']['tesseract_ocr']` - Default `false`. 

I want the packages to be usable on Windows, Mac, Linux (they contain the stuff for all systems - one pack for all).
With these options not selected, the setup process will only add system specific native stuff (Windows: support for
both Java 32-Bit and Java 64-Bit is added).

- `node['sikulix']['setup']['system_all']` - Default `false`. 

Other:

- `node['sikulix']['setup']['buildv']` - Build libVisionProxy.so. Default `false`.
- `node['sikulix']['setup']['notest']` - Skip validation after install. Default `false`.
- `node['sikulix']['setup']['clean']` - Clean directory. Default `false`.

## Getting Help

- Ask specific questions on [Stack Overflow](http://stackoverflow.com/questions/tagged/sikulix).
- Report bugs and discuss potential features in [Github issues](https://github.com/dhoer/chef-sikulix/issues).

## Contributing

Please refer to [CONTRIBUTING](https://github.com/dhoer/chef-sikulix/blob/master/CONTRIBUTING.md).

## License

MIT - see the accompanying [LICENSE](https://github.com/dhoer/chef-sikulix/blob/master/LICENSE.md) file for details.


CHANGELOG.md
# Changelog

## 3.1.0

- Support for Mac OS X

## 3.0.0

- Support for Sikulix 1.1.0 

## 2.0.3

- GitHub issues link invalid

## 2.0.2

- No X11 DISPLAY

## 2.0.1

- Change license from All Rights Reserved to MIT

## 2.0.0

- Add ubuntu support
- Remove support for JRuby Addons and Remote Server
- Refactor attribute names

## 1.0.0
  
- Add support for Remote Server
- Refactor Setup attribute names
- Place guard on Setup to run only once
- Fix memory bug with Setup

## 0.1.0

- Initial release using sikulix v1.1.0 nightly builds


spec/spec_helper.rb
# Added by ChefSpec
require 'chefspec'

ChefSpec::Coverage.start!


spec/unit/setup_spec.rb
require 'spec_helper'

describe 'sikulix::setup' do
  context 'windows' do
    let(:chef_run) do
      ChefSpec::Runner.new(platform: 'windows', version: '2008R2') do |node|
        node.set['java']['windows']['url'] = 'https://s3.amazonaws.com/prsnpublic/jdk-7u72-windows-i586.exe'
        node.set['sikulix']['setup']['java_api'] = true
      end.converge(described_recipe)
    end

    it 'creates sikulix home directory' do
      expect(chef_run).to create_directory('C:/sikulix')
    end

    it 'downloads sikulix setup' do
      expect(chef_run).to create_remote_file('C:/sikulix/sikulixsetup.jar').with(
        source: 'http://nightly.sikuli.de/sikulixsetup-1.1.0.jar'
      )
    end

    it 'executes sikulix setup with java_api option' do
      expect(chef_run).to run_execute(
        "\"C:/Windows/System32/java.exe\" -jar \"C:/sikulix/sikulixsetup.jar\" options 2"
      )
    end
  end

  context 'debian' do
    let(:chef_run) do
      ChefSpec::Runner.new(platform: 'ubuntu', version: '14.04') do |node|
        node.set['sikulix']['setup']['java_api'] = true
      end.converge(described_recipe)
    end

    it 'installs unzip package' do
      expect(chef_run).to install_package('unzip')
    end

    it 'installs libopencv-dev package' do
      expect(chef_run).to install_package('libopencv-dev')
    end

    it 'installs libtesseract-dev package' do
      expect(chef_run).to install_package('libtesseract-dev')
    end

    it 'creates sikulix home directory' do
      expect(chef_run).to create_directory('/usr/local/sikulix')
    end

    it 'creates sikulix libs directory' do
      expect(chef_run).to create_directory('/usr/local/sikulix/libs')
    end

    it 'downloads sikulix setup' do
      expect(chef_run).to create_remote_file('/usr/local/sikulix/sikulixsetup.jar').with(
        source: 'http://nightly.sikuli.de/sikulixsetup-1.1.0.jar'
      )
    end

    it 'executes sikulix setup with java_api option' do
      expect(chef_run).to run_execute(
        "\"/usr/bin/java\" -jar \"/usr/local/sikulix/sikulixsetup.jar\" options 2"
      )
    end
  end
end


spec/unit/default_spec.rb
require 'spec_helper'

describe 'sikulix_test::default' do
  before do
    stub_command('netsh advfirewall firewall show rule name=sikulixremoteserver > nul').and_return(false)
    allow(::File).to receive(:exist?) { false }
  end

  context 'windows' do
    let(:chef_run) do
      ChefSpec::SoloRunner.new(platform: 'windows', version: '2008R2') do |node|
        node.set['sikulix']['setup']['java_api'] = true
        node.set['sikulix']['windows']['home'] = 'C:/SikuliX'
        node.set['sikulix']['windows']['java'] = 'C:/java/bin/java.exe'
      end.converge(described_recipe)
    end

    it 'creates sikulix home directory' do
      expect(chef_run).to create_directory('C:/SikuliX')
    end

    it 'downloads sikulix setup' do
      expect(chef_run).to create_remote_file('C:/SikuliX/sikulixsetup-1.1.0.jar')
    end

    it 'executes sikulix setup with options' do
      expect(chef_run).to run_execute("\"C:/java/bin/java.exe\" -jar \"C:/SikuliX/sikulixsetup-1.1.0.jar\" options 2")
    end

    it 'creates SIKULIX_HOME' do
      expect(chef_run).to create_env('SIKULIX_HOME')
    end
  end

  context 'macosx' do
    let(:chef_run) do
      ChefSpec::SoloRunner.new(platform: 'mac_os_x', version: '10.10') do |node|
        node.set['sikulix']['setup']['java_api'] = true
      end.converge(described_recipe)
    end

    it 'creates sikulix home directory' do
      expect(chef_run).to create_directory('/opt/SikuliX')
    end

    it 'downloads sikulix setup' do
      expect(chef_run).to create_remote_file('/opt/SikuliX/sikulixsetup-1.1.0.jar')
        .with(source: 'https://launchpad.net/sikuli/sikulix/1.1.0/+download/sikulixsetup-1.1.0.jar')
    end

    it 'executes sikulix setup with java_api option' do
      expect(chef_run).to run_execute(
        "\"/usr/bin/java\" -jar \"/opt/SikuliX/sikulixsetup-1.1.0.jar\" options 2")
    end

    it 'executes chmod on home directory and copies SikuliX.app to Applications' do
      expect(chef_run).to run_execute('chmod -R 755 /opt/SikuliX && cp -R /opt/SikuliX/SikuliX.app /Applications')
    end
  end

  context 'ubuntu' do
    let(:chef_run) do
      ChefSpec::SoloRunner.new(platform: 'ubuntu', version: '14.04') do |node|
        node.set['sikulix']['setup']['java_api'] = true
      end.converge(described_recipe)
    end

    it 'installs wmctrl package' do
      expect(chef_run).to install_package('wmctrl')
    end

    it 'installs xdotool package' do
      expect(chef_run).to install_package('xdotool')
    end

    it 'installs libopencv-dev package' do
      expect(chef_run).to install_package('libopencv-dev')
    end

    it 'installs libtesseract-dev package' do
      expect(chef_run).to install_package('libtesseract-dev')
    end

    it 'creates sikulix home directory' do
      expect(chef_run).to create_directory('/opt/SikuliX')
    end

    it 'downloads sikulix setup' do
      expect(chef_run).to create_remote_file('/opt/SikuliX/sikulixsetup-1.1.0.jar')
        .with(source: 'https://launchpad.net/sikuli/sikulix/1.1.0/+download/sikulixsetup-1.1.0.jar')
    end

    it 'executes sikulix setup with java_api option' do
      expect(chef_run).to run_execute(
        "\"/usr/bin/java\" -jar \"/opt/SikuliX/sikulixsetup-1.1.0.jar\" options 2")
    end

    it 'executes chmod on home directory' do
      expect(chef_run).to run_execute('chmod -R 755 /opt/SikuliX')
    end
  end

  context 'no options' do
    let(:chef_run) do
      ChefSpec::SoloRunner.new(platform: 'ubuntu', version: '14.04').converge(described_recipe)
    end

    it 'fails' do
      expect { chef_run }.to raise_error(RuntimeError, 'SikuliX setup has no options selected!')
    end
  end
end


recipes/setup.rb
root = platform?('windows') ? node['sikulix']['windows']['root'] : node['sikulix']['linux']['root']
java = platform?('windows') ? node['sikulix']['windows']['java'] : node['sikulix']['linux']['java']

match = /.*\/(.*).jar/.match(node['sikulix']['src']['sikulixsetup']).captures[0]
v = /.*\-(\d|.*).jar/.match(node['sikulix']['src']['sikulixsetup']).captures[0]
dir = "#{root}/sikulix-#{v}"
link = "#{root}/sikulix"

directory dir do
  recursive true
  mode '0775'
  action :create
end

link link do
  to dir
end

node['sikulix']['packages']['debian'].each do |pkg|
  package(pkg)
end if platform_family?('debian')

remote_file "#{dir}/#{match}.jar" do
  source node['sikulix']['src']['sikulixsetup']
end

opts = []
opts << '1.1' if node['sikulix']['setup']['ide_jython'] == true
opts << '1.2' if node['sikulix']['setup']['ide_jruby'] == true
opts << '2' if node['sikulix']['setup']['java_api'] == true
opts << '3' if node['sikulix']['setup']['tesseract_ocr'] == true
opts << '4' if node['sikulix']['setup']['system_all'] == true
opts << '4.1' if node['sikulix']['setup']['system_windows'] == true
opts << '4.2' if node['sikulix']['setup']['system_mac'] == true
opts << '4.3' if node['sikulix']['setup']['system_linux'] == true
opts << 'buildv' if node['sikulix']['setup']['buildv'] == true
opts << 'notest' if node['sikulix']['setup']['notest'] == true
opts << 'clean' if node['sikulix']['setup']['clean'] == true

if opts.size == 0
  log 'SikuliX setup has no options selected - nothing to do' do
    level :warn
  end
else
  execute "\"#{java}\" -jar \"#{link}/#{match}.jar\" options #{opts.join(' ')}"
end


recipes/remoteserver.rb
home = platform?('windows') ? node['sikulix']['windows']['home'] : node['sikulix']['linux']['home']
java = platform?('windows') ? node['sikulix']['windows']['java'] : node['sikulix']['linux']['java']

if platform_family?('windows')
  directory "#{home}/bin"

  cmd_file = "#{home}/bin/sikulixremoteserver.cmd"

  file cmd_file do
    content "\"#{java}\" #{node['sikulix']['remoteserver']['jvm_args']} -jar "\
      "\"#{home}/sikulixremoteserver.jar\" #{node['sikulix']['remoteserver']['port']}"
    action :create
    not_if { ::File.exist?(cmd_file) }
    notifies :request, 'windows_reboot[Reboot to start sikulixremoteserver]' if platform?('windows')
  end

  startup = "C:\\Users\\#{node['sikulix']['username']}\\AppData\\Roaming\\Microsoft\\Windows"\
    '\\Start Menu\\Programs\\Startup\\sikulixremoteserver.lnk'

  windows_shortcut startup do
    target cmd_file
    cwd home
    action :create
  end

  autologin_values = [
    { name: 'AutoAdminLogon', type: :string, data: '1' },
    { name: 'DefaultUsername', type: :string, data: node['sikulix']['username'] },
    { name: 'DefaultPassword', type: :string, data: node['sikulix']['password'] }
  ]

  autologin_values << {
    name: 'DefaultDomainName', type: :string, data: node['sikulix']['domain']
  } unless node['sikulix']['domain'].nil?

  registry_key 'HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows NT\CurrentVersion\Winlogon' do
    values autologin_values
    action :create
  end

  execute "Firewall rule sikulixremoteserver for port #{node['sikulix']['remoteserver']['port']}" do
    command 'netsh advfirewall firewall add rule name=sikulixremoteserver protocol=TCP dir=in profile=any'\
            " localport=#{node['sikulix']['remoteserver']['port']} remoteip=any localip=any action=allow"
    action :run
    not_if 'netsh advfirewall firewall show rule name=sikulixremoteserver > nul'
  end

  include_recipe 'windows::reboot_handler'

  windows_reboot 'Reboot to start sikulixremoteserver' do
    reason 'Reboot to start sikulixremoteserver'
    timeout node['windows']['reboot_timeout']
    action :nothing
  end
elsif platform_family?('debian')
  template '/etc/init.d/sikulixremoteserver' do
    source 'debian.erb'
    mode '0755'
    variables(
      jar: "#{home}/sikulixremoteserver.jar",
      port: node['sikulix']['remoteserver']['port'],
      display: node['sikulix']['remoteserver']['display'],
      java: java,
      jvm_args: node['sikulix']['remoteserver']['jvm_args']
    )
    notifies(:restart, 'service[sikulixremoteserver]')
  end

  service 'sikulixremoteserver' do
    action [:enable, :start]
  end
else
  log 'SikuliX Remote Server cannot be installed on this platform.' do
    level :warn
  end
end


recipes/default.rb
if platform?('windows')
  home = node['sikulix']['windows']['home']
  java = node['sikulix']['windows']['java']
else
  home = node['sikulix']['unix']['home']
  java = node['sikulix']['unix']['java']
end

match = %r{.*/(.*).jar}.match(node['sikulix']['setup']['url']).captures[0]

directory home do
  recursive true
  mode '0775'
  action :create
end

node['sikulix']['packages']['debian'].each do |pkg|
  package(pkg)
end if platform_family?('debian')

remote_file "#{home}/#{match}.jar" do
  source node['sikulix']['setup']['url']
end

opts = []
opts << '1.1' if node['sikulix']['setup']['ide_jython']
opts << '1.2' if node['sikulix']['setup']['ide_jruby']
opts << '2' if node['sikulix']['setup']['java_api']
opts << '3' if node['sikulix']['setup']['tesseract_ocr']
opts << '4' if node['sikulix']['setup']['system_all']

opts << 'buildv' if node['sikulix']['setup']['buildv']
opts << 'not' if node['sikulix']['setup']['notest']
opts << 'clean' if node['sikulix']['setup']['clean']

if opts.size == 0
  fail('SikuliX setup has no options selected!')
else
  execute "\"#{java}\" -jar \"#{home}/#{match}.jar\" options #{opts.join(' ')}"
end

case node['platform']
when 'mac_os_x'
  execute "chmod -R 755 #{home} && cp -R #{home}/SikuliX.app /Applications"
when 'windows'
  env 'SIKULIX_HOME' do
    value home
  end
else # linux
  execute "chmod -R 755 #{home}"
end
